src/overload_hessian.jl

```diff
@@ -1,128 +1,99 @@
-## 1-to-1
-for fn in ops_1_to_1_s
-    @eval Base.$fn(t::HessianTracer) = promote_order(t)
-end
-for fn in ops_1_to_1_f
-    @eval Base.$fn(t::HessianTracer) = t
-end
-
-for fn in union(ops_1_to_1_z, ops_1_to_1_const)
-    @eval Base.$fn(::T) where {T<:HessianTracer} = empty(T)
-end
-
-## 2-to-1
-# Including second-order only
-for fn in ops_2_to_1_ssc
-    @eval function Base.$fn(a::HessianTracer, b::HessianTracer)
-        a = promote_order(a)
-        b = promote_order(b)
-        return distributive_merge(a, b)
+# TODO REVIEW: this keeps popping up so often that think it's worth an abstraction.  
+# It was previously called "promote_order".
+# The overly descriptive name is TEMPORARY.
+function tracer_from_hessian_OR_outer_product_OR_of_gradient(
+    t::T
+) where {T<:GlobalHessianTracer}
+    hessian_out = deepcopy(t.hessian)
+    hessian_out = outer_product_or!(hessian_out, t.gradient, t.gradient)
+    return T(t.gradient, hessian_out)
+end
+
+# ## 1-to-1
+for fn in ops_1_to_1
+    @eval function Base.$fn(t::T) where {T<:GlobalHessianTracer}
+        if is_seconder_zero_global($fn)
+            if is_firstder_zero_global($fn)
+                return empty(T)
+            else
+                return t
+            end
+        else
+            return tracer_from_hessian_OR_outer_product_OR_of_gradient(t)
+        end
     end
-    @eval Base.$fn(t::HessianTracer, ::Number) = promote_order(t)
-    @eval Base.$fn(::Number, t::HessianTracer) = promote_order(t)
 end
 
-for fn in ops_2_to_1_ssz
-    @eval function Base.$fn(a::HessianTracer, b::HessianTracer)
-        a = promote_order(a)
-        b = promote_order(b)
-        return additive_merge(a, b)
+# ## 2-to-1
+for fn in ops_2_to_1
+    @eval function Base.$fn(a::T, b::T) where {G,H,T<:GlobalHessianTracer{G,H}}
+        gradient_out, hessian_out = if !is_firstder_arg1_zero_global($fn)
+            if !is_firstder_arg2_zero_global($fn)
+                (a.gradient ∨ b.gradient, a.hessian ∨ b.hessian)
+            else
+                (a.gradient, deepcopy(a.hessian))
+            end
+        else # is_firstder_arg1_zero_global
+            if !is_firstder_arg2_zero_global($fn)
+                (b.gradient, deepcopy(b.hessian))
+            else
+                (empty_sparse_vector(G), empty_sparse_matrix(H))
```

Comment:
is there a need to specify `empty_sparse_vector` when `G` is already an abstract sparse vector?

Reply:
Some two-argument functions have {zero first- and second-order derivatives} on both input, e.g. `div`:
```
julia> div(2.1, 1.3)
1.0

julia> div(2.1, 1.33)
1.0

julia> div(2.12, 1.33)
1.0
```
The code above essentially creates an empty tracer. For comparison:
```Julia
function empty(::Type{GlobalHessianTracer{G,H}}) where {G,H}
    return GlobalHessianTracer{G,H}(empty_sparse_vector(G), empty_sparse_matrix(H))
end
```

Reply:
what I mean is that `empty(G)` is just as specific as `empty_sparse_vector(G)`

Reply:
Good point, will simplify